### PR TITLE
normalize execution summary (addresses #1260 and #1261)

### PR DESCRIPTION
**Pull request by demoray**
* allow configuration of the number of tasks to show in the execution summary
* always preface each task in the execution summary with the number of times the task was called

### CODE REVIEW

File: luigi/execution_summary.py [hunk 1]
```diff
@@ -89,34 +91,39 @@ def _get_str(task_dict, extra_indent):
     """
     This returns a string for each status
     """
+    summary_length = int(configuration.get_config().get('core', 'summary-length', 5))
```
Comment (Tarrasch) [inline luigi/execution_summary.py]: This is not how we do configuration in luigi. :)

http://luigi.readthedocs.org/en/latest/configuration.html#configuration-classes

File: luigi/execution_summary.py [hunk 2]
```diff
@@ -89,34 +91,39 @@ def _get_str(task_dict, extra_indent):
...
+        prefix_size = 4
         if extra_indent:
-            row += '    '
-        if len(lines) >= 5:
-            """
-            This is how many rows will be printed for each status. If you want fewer rows you can lower the limit.
-            """
-            row += '...'
-            lines.append(row)
+            prefix_size = 8
```
Comment (Tarrasch) [inline luigi/execution_summary.py]: I would just write `prefix_size = 8 if extra_indent else 4` :)

File: luigi/execution_summary.py [hunk 3]
```diff
@@ -126,15 +133,15 @@ def _get_str(task_dict, extra_indent):
...
                 if len(tasks) == 1:
-                    row += '- {0} {1}'.format(len(tasks), tasks[0])
+                    line = prefix + '- {0} {1}'.format(len(tasks), tasks[0])
                 if len(tasks) == 2:
-                    row += '- {0} and {1}'.format(tasks[0], tasks[1])
+                    line = prefix + '- {0} {1} and {2}'.format(len(tasks), tasks[0], tasks[1])
                 if len(tasks) > 2:
-                    row += '- {0} and {1} other {2}'.format(tasks[0], len(tasks) - 1, task_family)
-        lines.append(row)
+                    line = prefix + '- {0} {1} ...'.format(len(tasks), tasks[0])
```
Comment (Tarrasch) [inline luigi/execution_summary.py]: Hmm, Wasn't it nicer and shorter to just do `row += blah` instead of `line =  prefix + blah`. I don't mind the new variable name, but I think it looked ok with `+=`, why did you remove that?

File: test/execution_summary_test.py [hunk 4]
```diff
@@ -69,16 +69,90 @@ def requires(self):
...
+                    'This progress looks :( because there were failed tasks',
+                    '',
+                    '===== Luigi Execution Summary =====',
+                    '']
+        result = summary.split('\n')
+        self.assertEqual(len(result), len(expected))
+        for i, line in enumerate(result):
+            self.assertEqual(line, expected[i])
+
+    @with_config({'core': {'summary-length': '1'}})
```
Comment (Tarrasch) [inline test/execution_summary_test.py]: Yea, don't use section `[core]` for this. Maybe use `execution_summary`, and then the cli parameter will be something like`--execution-summary-length 3`. Right?

File: test/execution_summary_test.py [hunk 5]
```diff
@@ -616,21 +718,33 @@ def requires(self):
...
+                    '    * 1 were missing external dependencies:',
+                    '        - 1 MyExternal()',
+                    '    * 21 had missing external dependencies:',
+                    '        - 10 DateTask(date=1998-03-23...1998-04-01, num=5)',
+                    '        - 1 EntryPoint()',
+                    '        - 10 Foo(num=100, num2=12) ...',
+                    '',
+                    'This progress looks :| because there were missing external dependencies',
+                    '',
+                    '===== Luigi Execution Summary =====',
```
Comment (Tarrasch) [inline test/execution_summary_test.py]: Nice! :)

Comment (Tarrasch) [inline test/execution_summary_test.py]: Actually, an this be even dicer with:

```
dedent("""
Scheduled 218 tasks ...
* blah blah
...
""")
```

File: test/execution_summary_test.py [hunk 6]
```diff
@@ -700,3 +814,6 @@ class Foo(RunOnceTask):
         self.assertNotIn('The other workers were', s)
         self.assertIn('This progress looks :) because there were no failed ', s)
         self.assertNotIn('\n\n\n', s)
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (Tarrasch) [inline test/execution_summary_test.py]: Why? Just run `tox -e py27-nonhdfs test/execution_summary_test.py`?

File: doc/configuration.rst [hunk 7]
```diff
@@ -243,6 +243,9 @@ worker-wait-interval
   for another job after the scheduler has said that it does not have any
   available jobs.
 
+summary-length
+  Maxinum number of tasks to show in an execution summary.  If the value is 0,
+  then all tasks will be displayed.  Default value is 5.
```
Comment (Tarrasch) [inline doc/configuration.rst]: Thanks for also writing the docs :)

File: luigi/execution_summary.py (hunk 1 above)
Comment (demoray) [inline luigi/execution_summary.py]: Is it?  I can only find this in use by HDFS.  Every other instance I can find of configuration does it in a similar fashion.

From [notifications.py](https://github.com/spotify/luigi/blob/master/luigi/notifications.py#L45):
        return configuration.get_config().get('core', 'email-type', 'plain')

From [postgres.py](https://github.com/spotify/luigi/blob/master/luigi/postgres.py#L107)
    marker_table = luigi.configuration.get_config().get('postgres', 'marker-table', 'table_updates')

[interface.py](https://github.com/spotify/luigi/blob/master/luigi/interface.py#L156)
    if not configuration.get_config().getboolean(
            'core', 'no_configure_logging', False):

That said, I should have used 'getint' rather than int(... get_config.get(...))  An update will be coming shortly.

File: luigi/execution_summary.py (hunk 3 above)
Comment (demoray) [inline luigi/execution_summary.py]: Shorter, yes.

The same variable was being appended across 37 lines of code, which was more than would fit in one terminal.  It wasn't immediately clear as to what was appending to row and when.  My personal taste is that _get_msg could use additional refactoring, but this change made it easier to trace the content of 'line' through the flow.

File: test/execution_summary_test.py (hunk 4 above)
Comment (demoray) [inline test/execution_summary_test.py]: Alright.  Next update will adjust this.

File: test/execution_summary_test.py (hunk 5 above)
Comment (demoray) [inline test/execution_summary_test.py]: Alright.

File: luigi/execution_summary.py (hunk 1 above)
Comment (Tarrasch) [inline luigi/execution_summary.py]: They are wrong. I'll try to update them as I see it can be very confusing.

File: luigi/execution_summary.py (hunk 3 above)
Comment (Tarrasch) [inline luigi/execution_summary.py]: Ok. :)

**Comment by Tarrasch** [comment]
Can you also squash the commits please? I think it's enough with only 1 here. Unless you are willing to split them into logical commits (right now they are more of the "whoups, forgot XYZ" kind :)).

**Comment by demoray** [comment]
Squashed per request.

File: luigi/execution_summary.py [hunk 8]
```diff
@@ -89,34 +91,37 @@ def _get_str(task_dict, extra_indent):
     """
     This returns a string for each status
     """
+    summary_length = configuration.get_config().getint('execution_summary', 'summary-length', 5)
```
Comment (Tarrasch) [inline luigi/execution_summary.py]: This is still blocking merge. Please do it the luigi way as in the docs I linked to

File: doc/configuration.rst [hunk 9]
```diff
@@ -603,3 +603,13 @@ Parameters controlling storage of task history in a database
...
   sqlalchemy.
+
+
+[execution_summary]
+--------------
+
+Parameters controlling execution summary of a worker
+
+summary-length
+  Maxinum number of tasks to show in an execution summary.  If the value is 0,
```
Comment (Tarrasch) [inline doc/configuration.rst]: Maximum*

**Comment by Tarrasch** [comment]
This looks great, and thanks for making the summary more deterministic as well. But there just seem to be some small build issues blocking this from getting merged. Will you fix those so we can merge? :)

**Comment by Tarrasch** [comment]
Sorry for asking for another iteration. But can you please squash the commits? :)

This looks really great! Can't wait to get this merged!

**Comment by demoray** [comment]
Squashed again, per request.

**Comment by Tarrasch** [comment]
Thanks!!!
